pkg/labels/consul_applicator.go

```diff
@@ -188,5 +194,28 @@ func convertLabeledToKVP(l Labeled) (*api.KVPair, error) {
 	}, nil
 }
 
+// The current schema of labels in Consul is optimized for label retrieval on a single
+// object of a given ID. This layout is less effective when attempting to perform a
+// watch on the results of an arbitrary label selector, which is necessarily un-indexable.
+// This implementation will perform the simplest possible optimization, which is to cache
+// the entire contents of the tree under the given label type and share it with other watches.
+//
+// Due to the possibility that this tree might change quite frequently in environments
+// with lots of concurrent deployments, the aggregated result from Consul will only be queried
+// by a maximum frequency of once per LabelAggregationCap.
+//
+// Preparers should not use the consulApplicator's implementation of WatchMatches directly due
+// to the cost of querying for this subtree on any sizeable fleet of machines. Instead, preparers should
+// use the httpApplicator from a server that exposes the results of this (or another)
+// implementation's watch.
+func (c *consulApplicator) WatchMatches(selector labels.Selector, labelType Type, quitCh chan struct{}) chan *[]Labeled {
+	aggregator, ok := c.aggregators[labelType]
```

Comment:
You'll need to wrap the singleton lookup/instantiation code inside a mutex to prevent a data race.